sql/analyzer/resolve_database.go

```diff
@@ -11,53 +11,60 @@ func resolveDatabase(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error
 
 	a.Log("resolve database, node of type: %T", n)
 
-	switch v := n.(type) {
-	case *plan.ShowIndexes:
-		db, err := a.Catalog.Database(a.Catalog.CurrentDatabase())
-		if err != nil {
-			return nil, err
-		}
+	return n.TransformUp(func(n sql.Node) (sql.Node, error) {
+		switch v := n.(type) {
+		case *plan.ShowIndexes:
+			db, err := a.Catalog.Database(a.Catalog.CurrentDatabase())
+			if err != nil {
+				return nil, err
+			}
 
-		nc := *v
-		nc.Database = db
-		return &nc, nil
-	case *plan.ShowTables:
-		db, err := a.Catalog.Database(a.Catalog.CurrentDatabase())
-		if err != nil {
-			return nil, err
-		}
+			nc := *v
+			nc.Database = db
+			return &nc, nil
+		case *plan.ShowTables:
+			var dbName = v.Database.Name()
+			if dbName == "" {
+				dbName = a.Catalog.CurrentDatabase()
+			}
 
-		nc := *v
-		nc.Database = db
-		return &nc, nil
-	case *plan.CreateTable:
-		db, err := a.Catalog.Database(a.Catalog.CurrentDatabase())
-		if err != nil {
-			return nil, err
-		}
+			db, err := a.Catalog.Database(dbName)
+			if err != nil {
+				return nil, err
+			}
 
-		nc := *v
-		nc.Database = db
-		return &nc, nil
-	case *plan.Use:
-		db, err := a.Catalog.Database(v.Database.Name())
-		if err != nil {
-			return nil, err
-		}
+			nc := *v
+			nc.Database = db
+			return &nc, nil
+		case *plan.CreateTable:
+			db, err := a.Catalog.Database(a.Catalog.CurrentDatabase())
+			if err != nil {
+				return nil, err
+			}
 
-		nc := *v
-		nc.Database = db
-		return &nc, nil
-	case *plan.ShowCreateDatabase:
-		db, err := a.Catalog.Database(v.Database.Name())
-		if err != nil {
-			return nil, err
-		}
+			nc := *v
+			nc.Database = db
+			return &nc, nil
+		case *plan.Use:
+			db, err := a.Catalog.Database(v.Database.Name())
+			if err != nil {
+				return nil, err
+			}
+
+			nc := *v
+			nc.Database = db
+			return &nc, nil
+		case *plan.ShowCreateDatabase:
+			db, err := a.Catalog.Database(v.Database.Name())
+			if err != nil {
+				return nil, err
```

Comment:
This kind of block is almost in every case. Maybe it's good time to extract it to some function.

Reply:
But the type is different, so you can't extract it. We could extract the `db, err := a.Catalog.Database(v.Database.Name())`, but after error handling it would be the same amount of lines, so no point in extracting that either.

The only way I can imagine that we can use to extract this is create an interface that these nodes implement.

```go
type WithDatabaser interface {
  WithDatabase(sql.Database) (sql.Node, error)
}
```

Reply:
ohh, you're right - it sucks. Let's leave it.

Reply:
#generics 😜